### PR TITLE
Kill child processes when simexpal is terminated prematurely

### LINKED ISSUE #97: Child processes of simexpal keep running even when simexpal is terminated
While working on issue #81, I noticed that experiments started by simexpal keep running even after simexpal is terminated (prematurely).

We should consider catching those cases to terminate all child processes spawned by simexpal.

### PR DESCRIPTION
**Pull request by duylethanh**
This PR contains the following:

- fix #97

If ``SIGINT`` (CTRL + C) or ``SIGTERM`` is caught, kill running child processes before terminating. Note that we can not handle ``SIGKILL`` this way.

### CODE REVIEW

**Comment by avdgrinten** [comment]
I think we should integrate the signal handler into our event loop. This makes sure that all the code runs synchronously, without any surprises (e.g., we do not need to care *when* the signal handler runs).

We can do this by using the Python function `set_wakeup_fd`: create a pipe (`os.pipe()`), set the writer side as the wakeup FD, add the reader side to our `selector`. When the pipe becomes readable, read the signal number and terminate the process.

**Comment by duylethanh** [comment]
PR updated.

**Comment by duylethanh** [comment]
I added missing ``os.close()`` calls on the pipes.

File: simexpal/launch/common.py [hunk 1]
```diff
@@ -453,9 +466,16 @@ def close(self):
 		# Consume any output that might be ready.
 		events = sel.select(timeout=1)
 		for (sk, mask) in events:
+			if sk.fileobj == signal_reader:
+				child.kill()
+				sys.exit(1)
```
Comment (avdgrinten) [inline simexpal/launch/common.py:471]: Instead of calling `exit(1)`, I think we should follow the normal code path to clean up (e.g., to write the status file), and then we should re-raise the signal or raise an exception.

Comment (duylethanh) [reply simexpal/launch/common.py]: Done

**Comment by duylethanh** [comment]
PR updated

File: simexpal/launch/common.py [hunk 2]
```diff
@@ -453,9 +468,16 @@ def close(self):
 		# Consume any output that might be ready.
 		events = sel.select(timeout=1)
 		for (sk, mask) in events:
-			if not sk.data.progress():
+			if sk.fileobj == signal_reader:
+				sel.unregister(signal_reader)
+				child.kill()
+				do_exit = True
```
Comment (avdgrinten) [inline simexpal/launch/common.py:474]: Why is this needed? After the child is killed, our usual event loop logic should notice that the process terminated and proceed correctly, shouldn't it?

File: simexpal/launch/common.py [hunk 3]
```diff
@@ -453,9 +468,16 @@ def close(self):
 		# Consume any output that might be ready.
 		events = sel.select(timeout=1)
 		for (sk, mask) in events:
-			if not sk.data.progress():
+			if sk.fileobj == signal_reader:
+				sel.unregister(signal_reader)
```
Comment (avdgrinten) [inline simexpal/launch/common.py:472]: Why do we need to unregister `signal_reader` here? It probably makes more sense to just clear the status by reading from the pipe.

File: simexpal/queuesock.py [hunk 4]
```diff
@@ -50,6 +63,13 @@ def run(self):
 					connection = Connection(conn)
 
 					self.selector.register(conn, selectors.EVENT_READ, connection)
+				elif sk.fileobj == signal_reader:
+					if cur_subproc is not None:
+						cur_subproc.terminate()
```
Comment (avdgrinten) [inline simexpal/queuesock.py:68]: Why do we use terminate() and not kill() here (as in launchers/common.py)?

File: simexpal/queuesock.py [hunk 5]
```diff
@@ -50,6 +63,13 @@ def run(self):
...
+					os.close(signal_reader)
+					os.close(signal_writer)
+					self.close()
+					sys.exit(1)
```
Comment (avdgrinten) [inline simexpal/queuesock.py:72]: As in common.py, the usual event loop logic should handle this without any extra code.

File: simexpal/launch/common.py (hunk 2 above)
Comment (duylethanh) [reply simexpal/launch/common.py]: We also need to terminate the parent process. Otherwise experiments that where started through the same ``simex e launch`` command will be launched. I added a comment to this line to make it clear

File: simexpal/queuesock.py (hunk 4 above)
Comment (duylethanh) [reply simexpal/queuesock.py]: The subprocesses in ``queuesock.py`` are all simexpal processes that terminate accordingly when receiving ``SIGTERM`` or ``SIGINT``.
The subprocesses in ``common.py`` can be arbitrary processes, which could catch ``SIGTERM`` and ``SIGINT`` and keep runing. Thus, we use ``kill()`` in ``common.py``, which sends ``SIGKILL`` to make sure, we terminate the subprocess.

File: simexpal/queuesock.py (hunk 5 above)
Comment (duylethanh) [reply simexpal/queuesock.py]: Analogously to ``common.py`` we need to terminate the parent process. I changed up the handling of this case to avoid some duplicate code

File: simexpal/launch/common.py (hunk 3 above)
Comment (duylethanh) [reply simexpal/launch/common.py:497]: We have to unregister the ``signal_reader`` at some point, so I think after receiving the first signal, we no longer need to listen to it

**Comment by duylethanh** [comment]
PR updated and ready for review

**Comment by avdgrinten** [comment]
Ah, now I understand why we exit the program. But I still think the code would be cleaner if we let the event loops finish as usual (and clean up everything properly even in the exit path); afterwards, we can throw an exception to exit the whole program (instead of doing `sys.exit`). That way, higher level code can decide if it wants to exit or continue with the next run or so.

**Comment by duylethanh** [comment]
> Ah, now I understand why we exit the program. But I still think the code would be cleaner if we let the event loops finish as usual (and clean up everything properly even in the exit path); afterwards, we can throw an exception to exit the whole program (instead of doing `sys.exit`). That way, higher level code can decide if it wants to exit or continue with the next run or so.

Instead of ``sys.exit()`` we now raise the following error.

``` python
raise RuntimeError(
    "simexpal received a termination signal (either SIGINT or SIGTERM) and has terminated "
    "the current child process")
```

**Review by avdgrinten — APPROVED** [review]
LGTM.
